client/api.c

```diff
@@ -1904,6 +1904,353 @@ TDNFUpdateInfo(
+        {
+            dwError = ERROR_TDNF_INVALID_PARAMETER;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+    }
+    else if (pHistoryArgs->nCommand == HISTORY_CMD_UNDO ||
+             pHistoryArgs->nCommand == HISTORY_CMD_REDO)
+    {
+        if (pHistoryArgs->nFrom <= 1 || /* cannot undo or redo the base set */
+            pHistoryArgs->nFrom > pHistoryArgs->nTo)
+        {
+            dwError = ERROR_TDNF_INVALID_PARAMETER;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+    }
+    else if (pHistoryArgs->nCommand != HISTORY_CMD_INIT)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = TDNFRefresh(pTdnf);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    ts = rpmtsCreate();
+    if(!ts)
+    {
+        dwError = ERROR_TDNF_RPMTS_CREATE_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (rpmtsOpenDB(ts, O_RDONLY))
+    {
+        dwError = ERROR_TDNF_RPMTS_OPENDB_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if(rpmtsSetRootDir(ts, pTdnf->pArgs->pszInstallRoot))
+    {
+        dwError = ERROR_TDNF_RPMTS_BAD_ROOT_DIR;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = TDNFGetHistoryCtx(pTdnf, &ctx,
+                                pHistoryArgs->nCommand != HISTORY_CMD_INIT);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    rc = history_sync(ctx, ts);
+    if (rc != 0)
+    {
+        dwError = ERROR_TDNF_HISTORY_ERROR;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (pHistoryArgs->nCommand == HISTORY_CMD_INIT)
+    {
+        goto cleanup;
+    }
+    else if (pHistoryArgs->nCommand == HISTORY_CMD_ROLLBACK)
+    {
+        hd = history_get_delta(ctx, pHistoryArgs->nTo);
+    }
+    else if (pHistoryArgs->nCommand == HISTORY_CMD_UNDO)
+    {
+        hd = history_get_delta_range(ctx, pHistoryArgs->nFrom - 1, pHistoryArgs->nTo);
+    }
+    else if (pHistoryArgs->nCommand == HISTORY_CMD_REDO)
+    {
+        hd = history_get_delta_range(ctx, pHistoryArgs->nTo, pHistoryArgs->nFrom - 1);
+    }
+
+    if (hd == NULL)
+    {
+        dwError = ERROR_TDNF_HISTORY_ERROR;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (hd->added_count > 0)
+    {
+        dwError = TDNFAllocateMemory(
+                      hd->added_count+1, /* only added pkgs, plus a NULL ptr */
+                      sizeof(char*),
+                      (void**)&ppszPkgsNotResolved);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    hnm = history_nevra_map(ctx);
+    if (hnm == NULL)
+    {
+        dwError = ERROR_TDNF_HISTORY_ERROR;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    queue_init(&qInstall);
+    queue_init(&qErase);
+
+    for (int i = 0; i < hd->added_count; i++)
+    {
+        char *pszPkgName = history_get_nevra(hnm, hd->added_ids[i]);
+        if (pszPkgName)
+        {
+            Queue qResult = {0};
+            queue_init(&qResult);
+
+            dwError = SolvFindSolvablesByNevraStr(pTdnf->pSack->pPool,
```

Comment:
instead of 0 and 1 as installed can we put MACRO

Reply:
Done.